### PR TITLE
Create testing for abc_atlas_cache objects

### PR DESCRIPTION
**Pull request by morriscb**
Add testing for S3CloudCache object.
Add testing for LocalCache object.
Add testing for AbcProjectCache from s3 and local.

### CODE REVIEW

**Comment by morriscb** [comment]
Hey @danielsf I think I have a fare port of the unittesting from the SDK caches to here. Tried to clean things up a bit before sending it your way though my expectation of this review is that it will be pretty back and forth as you remember the previous testing and I try to implement what's missing, simplify/remove duplicate code.

File: tests/abc_atlas_cache/utils.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+import hashlib
+from moto import mock_aws
+import unittest
+import tempfile
+from pathlib import Path
+from abc_atlas_access.abc_atlas_cache.file_attributes import \
+    CacheFileAttributes
+
+
+def create_bucket(region_name: str,
```
Comment (danielsf) [inline tests/abc_atlas_cache/utils.py:12]: I'm being a little paranoid here, but can we make this a member of BaseCacheTestCase so no one accidentally calls this outside of a `@mock_aws` invocation.

Or else wrap this function in `@mock_aws` to be doubly-sure. When I first saw this function I did a double-take before realizing that it is only ever invoked in classes wrapped with `@mock_aws`

File: tests/abc_atlas_cache/test_cache.py [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+        bad_attribute = CacheFileAttributes(url='http://silly.url.com',
+                                            version=self.new_version,
+                                            file_size=1234,
+                                            local_path=self.cache_dir,
+                                            relative_path='junk.txt',
+                                            file_type='txt',
+                                            file_hash=true_checksum)
+        with pytest.raises(RuntimeError) as context:
+            cache._file_exists(bad_attribute)
+        assert 'but is not a file' in context.value.args[0]
```
Comment (danielsf) [inline tests/abc_atlas_cache/test_cache.py:165]: I suspect this was copy-and-pasted from SDK code written when *somebody* (cough, cough) wasn't totally comfortable with `pytest.raises`. Do you mind fixing this to use `pytest.raises(...., match='but is not a file')`

File: tests/abc_atlas_cache/test_cache.py [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+        cache._download_file(good_attributes)
+        assert expected_path.exists()
+        hasher = hashlib.md5()
+        with open(expected_path, 'rb') as in_file:
+            hasher.update(in_file.read())
+        assert hasher.hexdigest() == true_checksum
+
+    def test_download_data(self):
+        """
+        Test that S3CloudCache.download_data() correctly downloads files from
```
Comment (danielsf) [inline tests/abc_atlas_cache/test_cache.py:247]: This test no longer (or never did) downloads the data. It just checks that the `CacheFileAttributes` is correctly formulated (specifically, it asserts that the files it inquires about do not exist). Can you update the docstring?

File: tests/abc_atlas_cache/test_cache.py [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+
+        m_warn_type = 'OutdatedManifestWarning'
+
+        with pytest.warns(OutdatedManifestWarning) as warn:
+            cache.load_manifest(manifest_list[0])
+        ct = 0
+        for w in warn.list:
+            if w._category_name == m_warn_type:
+                msg = str(w.message)
+                assert 'is not the most up to date' in msg
```
Comment (danielsf) [inline tests/abc_atlas_cache/test_cache.py:335]: Can this also be done with `match=...` in the `pytest.warns()` context?

File: tests/abc_atlas_cache/test_cache.py [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+                assert 'S3CloudCache.compare_manifests' in msg
+                assert 'load_latest_manifest' in msg
+                ct += 1
+        assert ct > 0
+
+        # assert no warning is raised the second time by catching
+        # any warnings that are emitted and making sure they are
+        # not OutdatedManifestWarnings
```
Comment (danielsf) [inline tests/abc_atlas_cache/test_cache.py:343]: this comment is no longer accurate (I think *any* warning will cause the test to fail).

File: tests/abc_atlas_cache/test_cache.py [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+            out_file.write('babababa')
+        cache = S3CloudCache(self.cache_dir, self.test_bucket_name)
+        expected = f'Loading latest version -- {manifest_list[-1]}'
+        with pytest.warns(UserWarning, match=expected):
+            cache.load_last_manifest()
+        assert cache.current_manifest == manifest_list[-1]
+
+    def test_download_directory(self):
+        """
+        Test that S3CloudCache.download_data() correctly files in the directory
```
Comment (danielsf) [inline tests/abc_atlas_cache/test_cache.py:458]: Check the grammar on this docstring. I think it is supposed to be

"Test that S3CloudCache.download_data() correctly downloads files from S3"

File: tests/abc_atlas_cache/test_cache.py [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+        cache = S3CloudCache(self.cache_dir, self.test_bucket_name)
+
+        cache.load_manifest(self.new_manifest_string)
+
+        expected_data_path = (self.cache_dir / data_path)
+        assert not expected_data_path.exists()
+        expected_metadata_path = (self.cache_dir / metadata_path)
+        assert not expected_metadata_path.exists()
+
+        # test data_path
```
Comment (danielsf) [inline tests/abc_atlas_cache/test_cache.py:543]: Can you flesh out the comment (or docstring) to explain that, the difference between the `download_data` calls that fail and those that succeed is `skip_hash_check`?

**Review by danielsf — CHANGES_REQUESTED** [review]
Mostly some documentation requests.

There are a few cases where I ask that you clean up the SDK code you inherited (using `pytest.warns(...., match='...')` where I think it is possible. If I am wrong that this isn't possible (I know warnings are a pain), feel free to ignore.

**Review by danielsf — APPROVED** [review]
looks good

**Pull request merged**